src/BucketTesting/Validation/CampaignUtilizationValidator.php

```diff
@@ -0,0 +1,137 @@
+	const FEATURE_TOGGLE_METHOD_NAME = 'featureIsActive';
+
+	private $hasValidated = false;
+	private $campaignCollection;
+	private $errorLogger;
+	private $ignoredBuckets = [];
+
+	public function __construct( CampaignCollection $campaignCollection, array $ignoredBuckets, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->ignoredBuckets = $ignoredBuckets;
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate(): void {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		$featureToggleChecks = $this->getFeatureToggleChecks();
+		if ( $duplicates = $this->getDuplicates( $featureToggleChecks ) ) {
+			foreach ( $duplicates as $duplicate ) {
+				$this->errorLogger->addError(
+					'ChoiceFactory contains multiple checks for feature toggle: ' . $duplicate
+				);
+			}
+			return;
+		}
+
+		$configBucketIdList = $this->buildBucketIdsFromCampaignCollection(
+			$this->campaignCollection
+		);
+
+		/** Remove all buckets which were marked as test buckets in TEST_BUCKET_IGNORE_LIST */
+		$featureToggleChecks = array_filter( $featureToggleChecks, [ __CLASS__, 'filterBucketIds' ] );
+		$configBucketIdList = array_filter( $configBucketIdList, [ __CLASS__, 'filterBucketIds' ] );
+
+		/** Check if either the configuration files or the choice factory has rules not found in the other */
+		$codeConfigDiff = array_diff( $featureToggleChecks, $configBucketIdList );
+		$configCodeDiff = array_diff( $configBucketIdList, $featureToggleChecks );
+
+		if ( empty( $codeConfigDiff ) === false || empty( $configCodeDiff ) === false ) {
+			foreach ( $codeConfigDiff as $missingBucket ) {
+				$this->errorLogger->addError(
+					'Bucket ' . $missingBucket . ' is configured but no implementation can be found in ChoiceFactory.'
+				);
+			}
+			foreach ( $configCodeDiff as $featureToggleCheck ) {
+				$this->errorLogger->addError(
+					'Feature toggle check for ' . $featureToggleCheck . ' is implemented but no campaign configuration can be found.'
+				);
+			}
+		}
+
+		$this->hasValidated = true;
+	}
+
+	private function getFeatureToggleChecks(): array {
+		$featureToggleChecks = [];
+		foreach ( $this->parseMethodCalls() as $featureToggleCheck ) {
+			if ( count( $featureToggleCheck->args ) !== 1 ) {
+				throw new \LogicException( self::FEATURE_TOGGLE_METHOD_NAME . ' should have exactly one argument.' );
+			}
+			$featureToggleChecks[] = $featureToggleCheck->args[0]->value->value;
+		}
+		return $featureToggleChecks;
+	}
+
+	private function parseMethodCalls(): array {
+		$parser = ( new ParserFactory() )->create( ParserFactory::PREFER_PHP7 );
+		$nodeFinder = new NodeFinder();
+		$choiceFactoryCode = file_get_contents( self::CHOICE_FACTORY_LOCATION );
```

Comment:
Please inject a [`FileFetcher`](https://packagist.org/packages/jeroen/file-fetcher) implementation for better testability and error handling (exceptions instead of errors).

Reply:
That obviously gets rid of all my worries in a much nicer way, thanks.